### PR TITLE
Implement parsing query params

### LINKED ISSUE #3: Params for get request fail when proving a nested hash
```ruby
JSONAPI::Client.fetch(url, filter: { id: id })

# TypeError: Can't convert Hash into String.
# from /Users/pablocrivella/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/addressable-2.5.2/lib/addressable/uri.rb:370:in `rescue in encode_component'

```

### PR DESCRIPTION
**Pull request by pablocrivella**
Closes #3

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,10 @@
+root = true
```
[the remark this answers is not in the document]
Comment (pablocrivella) [reply .editorconfig]: Nope, this for anyone that clones the repo

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## [0.1.1] 2018-06-15
+
+### Added
+
+- Add support for parsing query params according to the spec.
```
[the remark this answers is not in the document]
Comment (pablocrivella) [reply CHANGELOG.md]: The `JSON:API` spec

File: bin/console [hunk 3]
```diff
@@ -5,5 +5,12 @@ require "bundler/setup"
 require "jsonapi/client"
 require "pry"
 require "pry-byebug"
+require "webmock"
+
+include WebMock::API # rubocop:disable Style/MixinUsage
+
+WebMock.enable!
+
+stub_request(:any, "https://api.example.com").to_return(body: { data: nil }.to_json)
```
[the remark this answers is not in the document]
Comment (pablocrivella) [reply bin/console]: I think its convenient for testing

File: lib/jsonapi/client/query_params.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+module JSONAPI
+  module Client
+    module QueryParams
+      class << self
+        def parse(params)
+          params.flatten.each_slice(2).with_object({}) do |(key, value), hash|
```
[the remark this answers is not in the document]
Comment (pablocrivella) [reply lib/jsonapi/client/query_params.rb]: You can check the tests for this

File: lib/jsonapi/client/query_params.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            hash.merge!(
+              value.is_a?(Hash) ? parse_nested_param(value, key) : { key => parse_param(value) }
+            )
+          end
+        end
+
+        private
+
+        def parse_nested_param(hash, prefix)
+          hash.map { |key, value| ["#{prefix}[#{key}]", parse_param(value)] }.to_h
```
[the remark this answers is not in the document]
Comment (pablocrivella) [reply lib/jsonapi/client/query_params.rb]: IMHO prefix is clear enough

File: lib/jsonapi/client/query_params.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        end
+
+        def parse_param(value)
+          case value
+          when Hash
+            raise ArgumentError, "Multi nested params are not supported."
```
[the remark this answers is not in the document]
Comment (pablocrivella) [reply lib/jsonapi/client/query_params.rb]: I don't think the spec supports it, i didn't find any example or use case where it should be supported.

File: lib/jsonapi/client/query_params.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (rjclaasen) [reply lib/jsonapi/client/query_params.rb]: Do you think that `params.flatten.each_slice(2).with_object({})` could be replaced by `params.each.with_object({})`?

File: spec/jsonapi/client/query_params_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    subject { JSONAPI::Client::QueryParams.parse(params) }
+
+    context "when params is an empty hash" do
+      let(:params) { {} }
+
+      it { is_expected.to eq({}) }
+    end
+
+    context "when params contains arrays" do
+      let(:params) { { include: %w[users users.comments] } }
```
[the remark this answers is not in the document]
Comment (pablocrivella) [reply spec/jsonapi/client/query_params_spec.rb]: I don't understand what would be the benefit of that change 😕

File: lib/jsonapi/client/query_params.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (rjclaasen) [reply lib/jsonapi/client/query_params.rb]: My main reason would be that `.flatten.each_slice(2).with_object({})` is a lot more difficult to understand than `.each.with_object({})`. We're also converting a hash to an array, then taking slices of size 2, which I thought does the same thing as just iterating over the hash.

When I ran some simple checks the results were identical, but the code we're working on might function differently.

```ruby
[1] pry(main)> h = { hi: "hi", hello: { hoi: "hoi", huhu: { ea: "ea" } } }
=> {:hi=>"hi", :hello=>{:hoi=>"hoi", :huhu=>{:ea=>"ea"}}}
[2] pry(main)> h.flatten.each_slice(2).with_object({}) { |(key, value), hash| puts "#{key}, #{value}, #{hash}" }
hi, hi, {}
hello, {:hoi=>"hoi", :huhu=>{:ea=>"ea"}}, {}
=> {}
[3] pry(main)> h.each.with_object({}) { |(key, value), hash| puts "#{key}, #{value}, #{hash}" }
hi, hi, {}
hello, {:hoi=>"hoi", :huhu=>{:ea=>"ea"}}, {}
=> {}
```

File: spec/jsonapi/client/query_params_spec.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (rjclaasen) [reply spec/jsonapi/client/query_params_spec.rb]: One less mental step to take for people looking at the code.

Not really important though, we can keep it as it is just fine :)

File: lib/jsonapi/client/query_params.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (pablocrivella) [reply lib/jsonapi/client/query_params.rb]: I'll give it a try, but i don't think that `each_with_object` is easier to understand to everyone, might be that for some devs the first line is easier to understand, it's more an opinion than a fact.

[the remark this answers is not in the document]
Comment (pablocrivella) [reply lib/jsonapi/client/query_params.rb]: It actually works the same! So it is indeed better with `each_with_object` 👍 nice suggestion

**Pull request merged**